### PR TITLE
Fix `specTpDistance` field in CSV export to Hastus

### PR DESCRIPTION
**Pull request by jarkkoka**
Do some logging improvements as well.

- - -
This change is [Reviewable](https://reviewable.io/reviews/HSLdevcom/jore4-hastus/20)

### CODE REVIEW

**Review by culka — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/HSLdevcom/jore4-hastus/20)* status: 0 of 5 files reviewed, 2 unresolved discussions (waiting on @jarkkoka)

___
*[`src/main/kotlin/fi/hsl/jore4/hastus/data/mapper/HastusConverter.kt` line 52 at r1](https://reviewable.io/reviews/HSLdevcom/jore4-hastus/20#-NWm3SCpCSXIHY8-dZkH:-NWm3SCpCSXIHY8-dZkI:bz793xe) ([raw file](https://github.com/HSLdevcom/jore4-hastus/blob/6f4e62a81523052b0ae38c29bd597fb4d79ab390/src/main/kotlin/fi/hsl/jore4/hastus/data/mapper/HastusConverter.kt#L52)):*
> ```Kotlin
>         routeIdAndVariant: String
>     ): List<RouteVariantPoint> {
>         var firstTimingPointEncountered = false
> ```

Unless I misremember, the first stop on a route should always be a timing point. It would be better to check the first stop and throw an exception if it's not a timing point.
___
*[`src/main/kotlin/fi/hsl/jore4/hastus/data/mapper/HastusConverter.kt` line 66 at r1](https://reviewable.io/reviews/HSLdevcom/jore4-hastus/20#-NWm4jAZBCMs87SfwrKb:-NWm4jA_9FZB1JHypARo:b-iyiim6) ([raw file](https://github.com/HSLdevcom/jore4-hastus/blob/6f4e62a81523052b0ae38c29bd597fb4d79ab390/src/main/kotlin/fi/hsl/jore4/hastus/data/mapper/HastusConverter.kt#L66)):*
> ```Kotlin
> 
>                 firstTimingPointEncountered = true
>                 accumulatedDistanceFromPreviousTimingPoint = it.distanceToNextStop
> ```

I don't really like mutating a var to keep track of the accumulated distance, but there's not any way to pair a value depending on previous value on the `.map` call.

**Review by jarkkoka — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/HSLdevcom/jore4-hastus/20)* status: 0 of 5 files reviewed, 2 unresolved discussions (waiting on @culka)

___
*[`src/main/kotlin/fi/hsl/jore4/hastus/data/mapper/HastusConverter.kt` line 52 at r1](https://reviewable.io/reviews/HSLdevcom/jore4-hastus/20#-NWm3SCpCSXIHY8-dZkH:-NWvwcOm-j3i0a-m-xrq:bkfvhmv) ([raw file](https://github.com/HSLdevcom/jore4-hastus/blob/6f4e62a81523052b0ae38c29bd597fb4d79ab390/src/main/kotlin/fi/hsl/jore4/hastus/data/mapper/HastusConverter.kt#L52)):*
<details><summary><i>Previously, culka (Teemu Mäkinen) wrote…</i></summary><blockquote>

Unless I misremember, the first stop on a route should always be a timing point. It would be better to check the first stop and throw an exception if it's not a timing point.
</blockquote></details>

Not sure whether that check should be done in the UI instead? E.g. how that exception should be translated to an HTTP status codes Would you like to pick one from the following status codes or something else (`200 Ok` with an error JSON)?
- `400 Bad Request`
- `404 Not Found`
- `406 Not Acceptable`
- `418 I'm a teapot`
- `422 Unprocessable Content`
- `500 Internal Server Error`
- `502 Bad Gateway`
- `503 Service Unavailable`
___
*[`src/main/kotlin/fi/hsl/jore4/hastus/data/mapper/HastusConverter.kt` line 66 at r1](https://reviewable.io/reviews/HSLdevcom/jore4-hastus/20#-NWm4jAZBCMs87SfwrKb:-NWvvlLB5_9ZwAnzb0x_:b-6og2ru) ([raw file](https://github.com/HSLdevcom/jore4-hastus/blob/6f4e62a81523052b0ae38c29bd597fb4d79ab390/src/main/kotlin/fi/hsl/jore4/hastus/data/mapper/HastusConverter.kt#L66)):*
<details><summary><i>Previously, culka (Teemu Mäkinen) wrote…</i></summary><blockquote>

I don't really like mutating a var to keep track of the accumulated distance, but there's not any way to pair a value depending on previous value on the `.map` call.
</blockquote></details>

I prefer mutability as well but cannot use e.g. the usual higher order functions such as `reduce` here. Any ideas? If I had come up with an easy immutable way (readable as well) I would have definitely used it.

**Review by jarkkoka — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/HSLdevcom/jore4-hastus/20)* status: 0 of 5 files reviewed, 2 unresolved discussions (waiting on @culka)

___
*[`src/main/kotlin/fi/hsl/jore4/hastus/data/mapper/HastusConverter.kt` line 52 at r1](https://reviewable.io/reviews/HSLdevcom/jore4-hastus/20#-NWm3SCpCSXIHY8-dZkH:-NWvyKFbCs6L8ZI35XCD:b-jblur2) ([raw file](https://github.com/HSLdevcom/jore4-hastus/blob/6f4e62a81523052b0ae38c29bd597fb4d79ab390/src/main/kotlin/fi/hsl/jore4/hastus/data/mapper/HastusConverter.kt#L52)):*
<details><summary><i>Previously, jarkkoka (Jarkko Kaura) wrote…</i></summary><blockquote>

Not sure whether that check should be done in the UI instead? E.g. how that exception should be translated to an HTTP status codes Would you like to pick one from the following status codes or something else (`200 Ok` with an error JSON)?
- `400 Bad Request`
- `404 Not Found`
- `406 Not Acceptable`
- `418 I'm a teapot`
- `422 Unprocessable Content`
- `500 Internal Server Error`
- `502 Bad Gateway`
- `503 Service Unavailable`
</blockquote></details>

So, your proposal therefore requires more detailed spec.

**Review by jarkkoka — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/HSLdevcom/jore4-hastus/20)* status: 0 of 5 files reviewed, 2 unresolved discussions (waiting on @culka)

___
*[`src/main/kotlin/fi/hsl/jore4/hastus/data/mapper/HastusConverter.kt` line 52 at r1](https://reviewable.io/reviews/HSLdevcom/jore4-hastus/20#-NWm3SCpCSXIHY8-dZkH:-NWw1AEA3Uy5B52ERmQS:b-y5jq1g) ([raw file](https://github.com/HSLdevcom/jore4-hastus/blob/6f4e62a81523052b0ae38c29bd597fb4d79ab390/src/main/kotlin/fi/hsl/jore4/hastus/data/mapper/HastusConverter.kt#L52)):*
<details><summary><i>Previously, jarkkoka (Jarkko Kaura) wrote…</i></summary><blockquote>

So, your proposal therefore requires more detailed spec.
</blockquote></details>

Actually, `200 Ok` with JSON error content cannot be used since the expected normal output is not JSON.

**Review by culka — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/HSLdevcom/jore4-hastus/20)* status: 0 of 5 files reviewed, 2 unresolved discussions (waiting on @jarkkoka)

___
*[`src/main/kotlin/fi/hsl/jore4/hastus/data/mapper/HastusConverter.kt` line 52 at r1](https://reviewable.io/reviews/HSLdevcom/jore4-hastus/20#-NWm3SCpCSXIHY8-dZkH:-NWwB1G4-DC2yhTdT3bA:bjhczwz) ([raw file](https://github.com/HSLdevcom/jore4-hastus/blob/6f4e62a81523052b0ae38c29bd597fb4d79ab390/src/main/kotlin/fi/hsl/jore4/hastus/data/mapper/HastusConverter.kt#L52)):*
<details><summary><i>Previously, jarkkoka (Jarkko Kaura) wrote…</i></summary><blockquote>

Actually, `200 Ok` with JSON error content cannot be used since the expected normal output is not JSON.
</blockquote></details>

`400 Bad Request` seems like the best response, since this is something the frontend can verify before sending the request. And it's different from a generic `500` which can result from pretty much any other failure.

**Review by jarkkoka — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/HSLdevcom/jore4-hastus/20)* status: 0 of 25 files reviewed, 2 unresolved discussions (waiting on @culka)

___
*[`src/main/kotlin/fi/hsl/jore4/hastus/data/mapper/HastusConverter.kt` line 52 at r1](https://reviewable.io/reviews/HSLdevcom/jore4-hastus/20#-NWm3SCpCSXIHY8-dZkH:-NXPs7JSCWAsIqGAcAj8:bo8t28z) ([raw file](https://github.com/HSLdevcom/jore4-hastus/blob/6f4e62a81523052b0ae38c29bd597fb4d79ab390/src/main/kotlin/fi/hsl/jore4/hastus/data/mapper/HastusConverter.kt#L52)):*
<details><summary><i>Previously, culka (Teemu Mäkinen) wrote…</i></summary><blockquote>

`400 Bad Request` seems like the best response, since this is something the frontend can verify before sending the request. And it's different from a generic `500` which can result from pretty much any other failure.
</blockquote></details>

This is now done but in a separate `ExportService` since validation should be kept separate from the conversion logic.

**Review by jarkkoka — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/HSLdevcom/jore4-hastus/20)* status: 0 of 25 files reviewed, 2 unresolved discussions (waiting on @culka)

___
*[`src/main/kotlin/fi/hsl/jore4/hastus/data/mapper/HastusConverter.kt` line 52 at r1](https://reviewable.io/reviews/HSLdevcom/jore4-hastus/20#-NWm3SCpCSXIHY8-dZkH:-NXPsNGnDXlCJs0yS4vp:b-e9tdhu) ([raw file](https://github.com/HSLdevcom/jore4-hastus/blob/6f4e62a81523052b0ae38c29bd597fb4d79ab390/src/main/kotlin/fi/hsl/jore4/hastus/data/mapper/HastusConverter.kt#L52)):*
<details><summary><i>Previously, jarkkoka (Jarkko Kaura) wrote…</i></summary><blockquote>

This is now done but in a separate `ExportService` since validation should be kept separate from the conversion logic.
</blockquote></details>

In other words, I prefer to follow [Separation of Concerns](https://en.wikipedia.org/wiki/Separation_of_concerns) principle.

**Review by jarkkoka — COMMENTED** [review]
I felt it necessary to do quite a lot of refactoring so that the validation fits well with the rest of the whole.

There was also a lot of room for improvement in the naming, so that the code would be more understandable and correspond to the common Jore4 naming conventions.

> *[Reviewable](https://reviewable.io/reviews/HSLdevcom/jore4-hastus/20#-:-NXPsfRgD0TlDiv7p8mw:b-ee20w7)* status: 0 of 25 files reviewed, 2 unresolved discussions (waiting on @culka)

**Review by jarkkoka — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/HSLdevcom/jore4-hastus/20)* status: 0 of 25 files reviewed, 3 unresolved discussions (waiting on @culka)

___
*[`src/main/kotlin/fi/hsl/jore4/hastus/graphql/GraphQLService.kt` line 10 at r1](https://reviewable.io/reviews/HSLdevcom/jore4-hastus/20#-NXPt6wI6twEPFlVP8oe:-NXPt6wI6twEPFlVP8of:b-tjdrnn) ([raw file](https://github.com/HSLdevcom/jore4-hastus/blob/6f4e62a81523052b0ae38c29bd597fb4d79ab390/src/main/kotlin/fi/hsl/jore4/hastus/graphql/GraphQLService.kt#L10)):*
> ```Kotlin
> import com.fasterxml.jackson.databind.ObjectMapper
> import fi.hsl.jore4.hastus.config.HasuraConfiguration
> import fi.hsl.jore4.hastus.data.hastus.IHastusData
> ```

As you can see from these import changes, Hastus data conversion is refactored out of this class as it does not belong to here.

**Review by jarkkoka — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/HSLdevcom/jore4-hastus/20)* status: 0 of 25 files reviewed, 6 unresolved discussions (waiting on @culka)

___
*[`src/main/kotlin/fi/hsl/jore4/hastus/data/mapper/ConversionsFromHastus.kt` line 25 at r2](https://reviewable.io/reviews/HSLdevcom/jore4-hastus/20#-NXPtuEtBxbOs0CZ5CfG:-NXPtuEtBxbOs0CZ5CfH:b-we7645) ([raw file](https://github.com/HSLdevcom/jore4-hastus/blob/b254906d95acb113bfdcd5745c4fa2604c110499/src/main/kotlin/fi/hsl/jore4/hastus/data/mapper/ConversionsFromHastus.kt#L25)):*
> ```Kotlin
> import kotlin.time.Duration.Companion.minutes
> 
> object JoreConverter {
> ```

I had a hard time figuring out from the previous class names which way the conversion works and between which domains (there are three: GraphQL, Jore4 and Hastus). That's why I felt it necessary to do all these renamings.
___
*[`src/main/kotlin/fi/hsl/jore4/hastus/graphql/converter/ConversionsFromGraphQL.kt` line 18 at r2](https://reviewable.io/reviews/HSLdevcom/jore4-hastus/20#-NXPtJJx2dzKcsac2fi9:-NXPtJJx2dzKcsac2fiA:b-we7645) ([raw file](https://github.com/HSLdevcom/jore4-hastus/blob/b254906d95acb113bfdcd5745c4fa2604c110499/src/main/kotlin/fi/hsl/jore4/hastus/graphql/converter/ConversionsFromGraphQL.kt#L18)):*
> ```Kotlin
> import kotlin.math.roundToInt
> 
> object ResultConverter {
> ```

I had a hard time figuring out from the previous class names which way the conversion works and between which domains (there are three: GraphQL, Jore4 and Hastus). That's why I felt it necessary to do all these renamings.
___
*[`src/main/kotlin/fi/hsl/jore4/hastus/graphql/converter/ConversionsToGraphQL.kt` line 27 at r2](https://reviewable.io/reviews/HSLdevcom/jore4-hastus/20#-NXPtlDc6PxFII8Q-Lje:-NXPtlDc6PxFII8Q-Ljf:b-we7645) ([raw file](https://github.com/HSLdevcom/jore4-hastus/blob/b254906d95acb113bfdcd5745c4fa2604c110499/src/main/kotlin/fi/hsl/jore4/hastus/graphql/converter/ConversionsToGraphQL.kt#L27)):*
> ```Kotlin
> import kotlin.time.toJavaDuration
> 
> object GraphQLConverter {
> ```

I had a hard time figuring out from the previous class names which way the conversion works and between which domains (there are three: GraphQL, Jore4 and Hastus). That's why I felt it necessary to do all these renamings.

**Review by jarkkoka — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/HSLdevcom/jore4-hastus/20)* status: 0 of 25 files reviewed, 7 unresolved discussions (waiting on @culka)

___
*[`src/main/resources/application.properties` line 5 at r2](https://reviewable.io/reviews/HSLdevcom/jore4-hastus/20#-NXPu5Kg5ZxXKsMY2a1r:-NXPu5Kg5ZxXKsMY2a1s:b-legiuq) ([raw file](https://github.com/HSLdevcom/jore4-hastus/blob/6ea204bc97c165160861567fdf64053da79de7f7/src/main/resources/application.properties#L5)):*
> ```INI
> 
> # Whether to return 400 (Bad Request) response if timing point validation fails.
> failOnTimingPointValidation=@failOnTimingPointValidation@
> ```

I decided to make timing point validation a configurable property in order to enable Hastus testing even with faulty data in the database.

**Review by culka — APPROVED** [review]
> Reviewed 1 of 5 files at r1, 24 of 24 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/HSLdevcom/jore4-hastus/20)* status: all files reviewed, 7 unresolved discussions (waiting on @jarkkoka)

___
*[`src/main/kotlin/fi/hsl/jore4/hastus/data/hastus/StopDistance.kt` line 9 at r2](https://reviewable.io/reviews/HSLdevcom/jore4-hastus/20#-NXz1MjwAe8ioa4NVfX-:-NXz1MjwAe8ioa4NVfX0:bhss71p) ([raw file](https://github.com/HSLdevcom/jore4-hastus/blob/6ea204bc97c165160861567fdf64053da79de7f7/src/main/kotlin/fi/hsl/jore4/hastus/data/hastus/StopDistance.kt#L9)):*
> ```Kotlin
>  * @property stopEnd Stop label from Jore4 fof the end
>  * @property baseInService Constant 1
>  * @property trimmedDistance The distance between the stops, no leading zeroes, format mmmm
> ```

This field is called `itn_edited_distance` in Hastus and in the documentation https://drive.google.com/drive/folders/1sCnrorUi-Y91axZ_GnR1sy9fac3hF8Dy , so maybe it should be kept as such.

**Review by jarkkoka — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/HSLdevcom/jore4-hastus/20)* status: all files reviewed, 7 unresolved discussions (waiting on @culka)

___
*[`src/main/kotlin/fi/hsl/jore4/hastus/data/hastus/StopDistance.kt` line 9 at r2](https://reviewable.io/reviews/HSLdevcom/jore4-hastus/20#-NXz1MjwAe8ioa4NVfX-:-NXzQCguCLgAlQn2kEcD:b-4uy45b) ([raw file](https://github.com/HSLdevcom/jore4-hastus/blob/6ea204bc97c165160861567fdf64053da79de7f7/src/main/kotlin/fi/hsl/jore4/hastus/data/hastus/StopDistance.kt#L9)):*
<details><summary><i>Previously, culka (Teemu Mäkinen) wrote…</i></summary><blockquote>

This field is called `itn_edited_distance` in Hastus and in the documentation https://drive.google.com/drive/folders/1sCnrorUi-Y91axZ_GnR1sy9fac3hF8Dy , so maybe it should be kept as such.
</blockquote></details>

Ok, thanks for clarification! I'll leave this out.
